core/module.go

```diff
@@ -4,66 +4,568 @@ import (
 }
 
-// BaseDigest gives a digest after unsetting Objects/Interfaces, which is useful
-// as a digest of the "base" Module that's stable before+after loading TypeDefs
-func (mod *Module) BaseDigest() (digest.Digest, error) {
-	mod = mod.Clone()
-	mod.Objects = nil
-	mod.Interfaces = nil
-	return stableDigest(mod)
+var _ Mod = (*Module)(nil)
+
+func (mod *Module) Name() string {
+	return mod.NameField
+}
+
+func (mod *Module) Dependencies() []Mod {
+	mods := make([]Mod, len(mod.DependenciesField))
+	for i, dep := range mod.DependenciesField {
+		mods[i] = dep.Self
+	}
+	return mods
+}
+
+func (mod *Module) Initialize(ctx context.Context, oldSelf dagql.Instance[*Module], newID *idproto.ID) (*Module, error) {
+	// construct a special function with no object or function name, which tells
+	// the SDK to return the module's definition (in terms of objects, fields and
+	// functions)
+	getModDefFn, err := newModFunction(
+		ctx,
+		mod.Query,
+		oldSelf.Self,
+		oldSelf.ID(),
+		nil,
+		mod.Runtime,
+		NewFunction("", &TypeDef{
+			Kind:     TypeDefKindObject,
+			AsObject: dagql.NonNull(NewObjectTypeDef("Module", "")),
+		}))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create module definition function for module %q: %w", mod.Name(), err)
+	}
+	result, err := getModDefFn.Call(ctx, newID, &CallOpts{Cache: true, SkipSelfSchema: true})
+	if err != nil {
+		return nil, fmt.Errorf("failed to call module %q to get functions: %w", mod.Name(), err)
+	}
+	inst, ok := result.(dagql.Instance[*Module])
+	if !ok {
+		return nil, fmt.Errorf("expected Module result, got %T", result)
+	}
+	newMod := inst.Self.Clone()
+	newMod.InstanceID = newID
+	return newMod, nil
+}
+
+func (mod *Module) Install(ctx context.Context, dag *dagql.Server) error {
+	slog.Debug("installing module", "name", mod.Name())
+	start := time.Now()
+	defer func() { slog.Debug("done installing module", "name", mod.Name(), "took", time.Since(start)) }()
+
+	for _, def := range mod.ObjectDefs {
+		objDef := def.AsObject.Value
+
+		slog.Debug("installing object", "name", mod.Name(), "object", objDef.Name)
+
+		// check whether this is a pre-existing object from a dependency module
+		modType, ok, err := mod.Deps.ModTypeFor(ctx, def)
+		if err != nil {
+			return fmt.Errorf("failed to get mod type for type def: %w", err)
+		}
+
+		if ok {
+			// NB: this is defense-in-depth to prevent SDKs or some other future
+			// component from allowing modules to extend external objects.
+
+			slog.Warn("found existing module type",
+				"name", mod.Name(),
+				"object", objDef.Name,
+				"sourceSame", modType.SourceMod() == mod,
+				"definesFields", len(objDef.Fields) > 0,
+				"definesFunctions", len(objDef.Functions) > 0,
+			)
+
+			if modType.SourceMod() != mod {
+				// modules can reference types from core/other modules as types, but they
+				// can't attach any new fields or functions to them
+				//
+				// TODO: what does "referencing" mean in this context? do we even get
```

Comment:
Know we talked about this previously, but looking at the concrete implementation now I think we actually can remove the whole `if modType.SourceMod() != mod {` block and just make `if ok {` return an error unconditionally.

When I last thought about this, I was thinking this chunk of code got called for every type that makes an appearance in the typedefs, even if only referenced by name as a return type/arg type. 

I'm pretty sure older incarnations of this code block actually did get called like that, but that was several refactors ago and it's now vestigial and can be removed. This is only called for the "top-level" objects newly defined by the module, so if anything tries to define an already existing type, that's always an error. Sorry for the wild goose chasing I may have set you on 😓

Reply:
> Know we talked about this previously, but looking at the concrete implementation now I think we actually can remove the whole `if modType.SourceMod() != mod {` block and just make `if ok {` return an error unconditionally.

A ha, that's exactly what I was wondering, glad you found this comment. 

> I'm pretty sure older incarnations of this code block actually did get called like that, but that was several refactors ago and it's now vestigial and can be removed. This is only called for the "top-level" objects newly defined by the module, so if anything tries to define an already existing type, that's always an error. Sorry for the wild goose chasing I may have set you on 😓

Yeah, I _think_ this predates namespacing, which should make conflicts really obtuse (technically still possible since we don't use a forbidden separator like `_` I think). No worries, I looked at this early on but squinted and moved on until the very end, when I just slapped a comment on it. :P